### PR TITLE
GCP: Allow to select algo when generating temporary SSH keypair

### PR DESCRIPTION
**Pull request by azr**
This is based on top of #10101, to solve the GCP case of #10074 this:

* adds  temporary_key_pair_type/temporary_key_pair_bits option to the communicator.SSH struct
* makes the GCP builder use those options along with the sshkey Package to generate the correct temporary keypair when needed.
* updates GCP docs + generate all code

### CODE REVIEW

File: builder/googlecompute/builder.go [hunk 1]
```diff
@@ -59,10 +59,15 @@ func (b *Builder) Run(ctx context.Context, ui packer.Ui, hook packer.Hook) (pack
...
 		new(StepCheckExistingImage),
-		&StepCreateSSHKey{
-			Debug:        b.config.PackerDebug,
-			DebugKeyPath: fmt.Sprintf("gce_%s.pem", b.config.PackerBuildName),
+		&communicator.StepSSHKeyGen{
+			CommConf:            &b.config.Comm,
+			SSHTemporaryKeyPair: b.config.Comm.SSH.SSHTemporaryKeyPair,
 		},
+		multistep.If(b.config.PackerDebug,
+			&communicator.StepDumpSSHKey{
```
Comment (SwampDragons) [inline builder/googlecompute/builder.go:67]: ooooh shiny.

**Comment by SwampDragons** [comment]
Tried to run it and it crashed on me -- 

```json
{
  "variables": {
      "image_name": "ubuntu1804-jenkins-{{isotime}}"
  },
  "builders": [
    {
      "type": "googlecompute",
      "image_name": "{{ user `image_name` | clean_resource_name}}",
      "account_file": "/myaccountfile.json",
      "project_id": "gifted-station-178619",
      "source_image_family": "ubuntu-1804-lts",
      "ssh_username": "packer",
      "zone": "us-central1-c",
      "state_timeout": "20m",
      "startup_script_file": "./shell_local_dummy.sh",
      "tags": ["packer"]
    }
  ],
  "provisioners": [
    {
      "pause_before": "10s",
      "type": "shell",
      "inline": ["echo {{ build `ID` }}"]
    }
  ]
}

```

**Comment by azr** [comment]
Weird, your file works on my machine 🤔  any special options ? Do you have the crash log ?

**Comment by azr** [comment]
NVM found it !

**Comment by azr** [comment]
Here's a testfile that works with any value:

```go
{
    "builders": [
        {
            "type": "googlecompute",
            "project_id": "azr-hashicorp",
            "image_name": "base-centos-7-{{timestamp}}",
            "source_image": "centos-7-v20200521",
            "source_image_family": "centos-7",
            "ssh_username": "centos",
            "temporary_key_pair_type": "ecdsa",
            "temporary_key_pair_bits": 521,
            "zone": "europe-west2-a",
            "machine_type": "n1-standard-1",
            "disk_size": 20,
            "enable_secure_boot": true,
            "enable_vtpm": true,
            "enable_integrity_monitoring": true
        }
    ],
    "provisioners": [
        {
            "destination": "/tmp/hm hi,\\ I have a space \u0026 ☺",
            "source": "log.go",
            "type": "file"
        }
    ],
    "variables": {
        "zone": "europe-west4-a"
    }
}
```

**Comment by azr** [comment]
Here's a cool HCL2 test file:

```hcl

variable "zone" {
  type    = string
  default = "europe-west4-a"
}

locals { timestamp = regex_replace(timestamp(), "[- TZ:]", "") }

source "googlecompute" "centos" {
  disk_size                   = 20
  enable_integrity_monitoring = true
  enable_secure_boot          = true
  enable_vtpm                 = true
  machine_type                = "n1-standard-1"
  project_id                  = "azr-hashicorp"
  source_image                = "centos-7-v20200521"
  source_image_family         = "centos-7"
  ssh_username                = "centos"
  zone                        = "europe-west2-a"
}

build {
  source "googlecompute.centos" {
    temporary_key_pair_type = "dsa"
    image_name              = "base-centos-7-dsa"
  }
  source "googlecompute.centos" {
    temporary_key_pair_type = "rsa"
    image_name              = "base-centos-7-rsa"
  }
  source "googlecompute.centos" {
    temporary_key_pair_type = "ecdsa"
    image_name              = "base-centos-7-ecdsa"
  }
  source "googlecompute.centos" {
    temporary_key_pair_type = "ed25519"
    image_name              = "base-centos-7-ed25519"
  }

  provisioner "file" {
    destination = "/tmp/hm hi,\\ I have a space & ☺"
    source      = "log.go"
  }
}

```

**Comment by SwampDragons** [comment]
cool! Your example works, but my original example still doesn't, and it does work on the master branch so we need to make sure that the default case still works.

**Comment by azr** [comment]
Ah indeed ! Thanks !

**Comment by azr** [comment]
Okay found the why, with no parameter this changes the default keytype from rsa to dsa ( because this was the default for `ssh-keygen` too and I thought this was the safest approach and that it would not break anything), this is clearly a breaking change so I'm making the default type `rsa` again.

File: helper/multistep/if.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package multistep
+
+// if returns step only if on is true.
+func If(on bool, step Step) Step {
```
Comment (nywilken) [inline helper/multistep/if.go:4]: Nice

**Review by nywilken — APPROVED** [review]
This looks great. Lots of nice little add-ons here. I have not tested against the examples but code wise it looks good. I approve, but will defer until @SwampDragons has given the green light that their use case has been satisfied.

**Comment by SwampDragons** [comment]
yep, works now!

**Pull request merged**